e2e/service_test.go

```diff
@@ -0,0 +1,197 @@
+	switch ts.ServiceType {
+	case serviceTypeEVM:
+		return getEVMVegetaTargets(ts, methods, gatewayURL)
+	case serviceTypeSolana:
+		return getSolanaVegetaTargets(ts, methods, gatewayURL)
+	case serviceTypeCometBFT:
+		return getCometBFTVegetaTargets(ts, methods, gatewayURL)
+	}
+	return nil, fmt.Errorf("unsupported service type: %s", ts.ServiceType)
+}
+
+func getExpectedID(serviceType serviceType) jsonrpc.ID {
+	switch serviceType {
+	case serviceTypeEVM:
+		return evmExpectedID
+	case serviceTypeSolana:
+		return solanaExpectedID
+	case serviceTypeCometBFT:
+		return cometbftExpectedID
+	default:
+		return jsonrpc.IDFromInt(1)
+	}
+}
+
+// -----------------------------------------------------------------------------
+// Utility Functions
+// -----------------------------------------------------------------------------
+
+// getRequestHeaders returns the HTTP headers for a given service ID, including Portal credentials if in load test mode.
+func getRequestHeaders(serviceID protocol.ServiceID) http.Header {
+	headers := http.Header{
+		"Content-Type":                    []string{"application/json"},
+		request.HTTPHeaderTargetServiceID: []string{string(serviceID)},
+	}
+
+	if cfg.getTestMode() == testModeLoad {
+		// Portal App ID is required for load tests
+		headers.Set(gateway.HttpHeaderPortalAppID, cfg.E2ELoadTestConfig.LoadTestConfig.PortalApplicationID)
+
+		// Portal API Key is optional for load tests
+		if cfg.E2ELoadTestConfig.LoadTestConfig.PortalAPIKey != "" {
+			headers.Set(gateway.HttpHeaderAuthorization, cfg.E2ELoadTestConfig.LoadTestConfig.PortalAPIKey)
+		}
+	}
+
+	return headers
+}
+
+// setServiceIDInGatewayURLSubdomain inserts the service ID as a subdomain in the gateway URL.
+//   - https://rpc.grove.city/v1 → https://F00C.rpc.grove.city/v1
+//   - http://localhost:3091/v1 → http://F00C.localhost:3091/v1
+//   - https://api.example.com/path?query=param → https://F00C.api.example.com/path?query=param
+//
+// TODO_TECHDEBT(@commoddity): Remove this once PATH in production supports service in headers
+//   - Issue: https://github.com/buildwithgrove/infrastructure/issues/91
+func setServiceIDInGatewayURLSubdomain(gatewayURL string, serviceID protocol.ServiceID) string {
+	parsedURL, err := url.Parse(gatewayURL)
+	if err != nil {
+		// If parsing fails, fall back to simple string insertion
+		return gatewayURL
+	}
+	parsedURL.Host = fmt.Sprintf("%s.%s", serviceID, parsedURL.Host)
+	return parsedURL.String()
+}
+
+// validate validates all test services
+func (ts *TestServices) validate() error {
+	if len(ts.Services) == 0 {
+		return fmt.Errorf("no test services configured")
+	}
+
+	for i, service := range ts.Services {
+		if err := ts.validateTestService(service, i); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validateTestService validates an individual test service and its config
+func (ts *TestServices) validateTestService(tc TestService, index int) error {
```

Comment:
Not a blocker: Should not be needed if the QoS implementation is used as the authority on the correct configuration for testing a service.

Reply:
Agreed. Will be captured in an issue.